CIP-0068 | Support multi-asset metadata a la CIP-0025

thaddeusdiamond opened the pull request:
Add version 4 metadata support with 721-ERC-style mappings to NFT (222), FT (333) and RFT (444) standards, ensuring consistency across all asset classes.

Changes:
- Add metadata_field union type supporting both direct and 721-map formats
- Add __RESERVE_KEYWORD_721_V4__ marker for 721-style detection
- Update retrieval steps to handle both metadata formats
- Support version 4 in FT standard
- Support version 3/4 in RFT standard
- Add 721-style JSON example for NFT standard
- Update pattern descriptions with asset_name terminology
- Fix URI support to include [* bounded_bytes] for RFT

thaddeusdiamond committed:
CIP-0068 | Support multi-asset metadata a la CIP-0025 (#1112)

Add version 4 metadata support with 721-ERC-style mappings to NFT (222),
FT (333) and RFT (444) standards, ensuring consistency across
all asset classes.

Changes:
- Add metadata_field union type supporting both direct and 721-map formats
- Add __RESERVE_KEYWORD_721_V4__ marker for 721-style detection
- Update retrieval steps to handle both metadata formats
- Support version 4 in FT standard
- Support version 3/4 in RFT standard
- Add 721-style JSON example for NFT standard
- Update pattern descriptions with asset_name terminology
- Fix URI support to include [* bounded_bytes] for RFT

rphair commented:
@thaddeusdiamond I was answering your comments in chronological order so I posted this response without seeing you've already done the proposed update:
* https://github.com/cardano-foundation/CIPs/issues/1111#issuecomment-3495360543

I think these changes are consistent with your suggestion of a new best-practice approach, the editorial guidelines, and the requirements of CIP-0068... but we'll follow the related issue in case other feedback indicates that a different approach should be taken.  In any case we'll tag this `Triage` so this can all be introduced at the next CIP meeting together (to include any review or resolutions in the meantime): https://hackmd.io/@cip-editors/123

thaddeusdiamond commented:
I think @Crypto2099 was onboard as is since it's backwards compatible but he can comment here.  Will close #1111

SmaugPool commented:
I'm not sure to understand the proposal. As the reference token asset_name must match the user token one modulo the asset_name_label, and both should have the same policy, is the plan to have several reference token in the same tx output and one datum?

In this case could you detail how much this saves if at all with examples? Because the additional reserved bool, `policy_id` and `asset_name` take space that was not required before. So I'm not convinced by the gain compared to several tx outputs, each with its own datum.

Overall please detail the aim of this proposal. What issue does it solve? I'm not sure consistency has been an issue. If it is, could you provide examples?

SmaugPool commented on CIP-0068/README.md:
```diff
@@ -188,6 +192,26 @@ metadata =
+      }
+    }
+  }
+
```
This is not some valid CDDL. It looks like you are confusing CDDL with JSON.
More importantly, this does not tell how the `policy_id` and `asset_name` keys are encoded (I guess byte encoding like CIP25v2 and all CIP68 fields?).

Overall we should avoid talking about JSON as metadata and datum are CBOR encoded and CBOR -> JSON is not bijective. This only leads to confusion. 

Note that JSON in CIP68 was only discussed once in an example of a process to convert datum to JSON.  And even this is slightly questionnable.

thaddeusdiamond replied:
Fixed. The CDDL now properly uses bounded_bytes for all keys:
- "721" key is encoded as bounded_bytes (UTF-8 bytes: 0x373231)
- policy_id keys are bounded_bytes (as in CIP-0025 version 2)
- asset_name keys are bounded_bytes (as in CIP-0025 version 2)

This matches CIP-0025 version 2 conventions where keys are byte-encoded rather than text strings.

rphair replied:
resolving for reasons noted & apparently fixed on one of the later commits.   @thaddeusdiamond please get in the habit of marking these "resolved" when you believe them to be fixed (reviewers e.g. @SmaugPool can/should also do this).

SmaugPool commented on CIP-0068/README.md:
````diff
@@ -236,6 +260,45 @@ Example datum as JSON:
 }
 ```
 
+Example datum as JSON (721-style):
````
Again what "721-style" means?

thaddeusdiamond commented:
> > I would think this would be up to the dApp developer? So long as it conforms to the spec they can either "respend" the datum and rewrite every related token metadata, or they could split it overall.
> 
> I'm not sure there is any choice: assuming there are n tokens in a single tx output with their metadata in the output datum. To update only one of the tokens metadata, you will have to spend all of them, at least as change, but CIP68 metadata lookup is described as:
> 
> > Construct reference NFT from user token
> > Look up reference NFT and find the output it's locked in.
> > Get the datum from the output and lookup metadata by going into the first field of constructor 0.
> 
> Note that the lookup only looks at the unspent output (not a previous output). Therefore you will have to rewrite the metadata of all tokens in a datum, hence it makes no sense to split them, it would be even costlier.

Makes sense.  So rewrite it then makes sense to me.

SmaugPool commented:
> > > I would think this would be up to the dApp developer? So long as it conforms to the spec they can either "respend" the datum and rewrite every related token metadata, or they could split it overall.
> > 
> > 
> > I'm not sure there is any choice: assuming there are n tokens in a single tx output with their metadata in the output datum. To update only one of the tokens metadata, you will have to spend all of them, at least as change, but CIP68 metadata lookup is described as:
> > > Construct reference NFT from user token
> > > Look up reference NFT and find the output it's locked in.
> > > Get the datum from the output and lookup metadata by going into the first field of constructor 0.
> > 
> > 
> > Note that the lookup only looks at the unspent output (not a previous output). Therefore you will have to rewrite the metadata of all tokens in a datum, hence it makes no sense to split them, it would be even costlier.
> 
> Makes sense. So rewrite it then makes sense to me.

This is quite a significant drawback compared to current CIP68 and even CIP25 (as only one of the tokens can be minted again to update its metadata), so maybe it's worth thinking more about it to avoid introducing a new flaw?

thaddeusdiamond commented on CIP-0068/README.md:
```diff
@@ -188,6 +192,26 @@ metadata =
+  {
+    __RESERVE_KEYWORD_721_V4__ : bool, ; Long name to avoid conflict
+
+    "721":
```
Removed the `__RESERVE_KEYWORD_721_V4__` marker. You're correct that the presence of the "721" key is sufficient to detect the nested map format. The format can be determined simply by checking if the metadata map contains a "721" key.

thaddeusdiamond commented on CIP-0068/README.md:
```diff
@@ -430,19 +523,39 @@ metadata =
+      }
+    }
+  }
+
```
All metadata_map definitions across 222, 333, and 444 standards have been updated with proper CDDL syntax using bounded_bytes for all keys, consistent with CIP-0025 version 2.

rphair replied:
(continued: also resolving)

thaddeusdiamond commented on CIP-0068/README.md:
```diff
@@ -578,6 +694,10 @@ versions of the affected tokens. `asset_name_labels` **MUST** only be marked obs
 
+#### version 4
+
+- Add backwards-compatible support for multi-asset 721-ERC-style metadata mappings
```
Clarified:
- "Backwards-compatible" means compatible with versions 1-3, which only supported direct metadata format. Version 4 adds the nested map format as an additional option.
- Removed "721-ERC-style" terminology entirely, replacing it with "nested map format following CIP-0025's policy_id -> asset_name structure"

rphair replied:
Thanks @SmaugPool & @thaddeusdiamond for working out these details & making a stronger spec & better written proposal: more uniquely qualifying Cardano for this type of application.  Based on this rigorous review I feel confident moving to a last round of review before merge.

thaddeusdiamond committed:
CIP-0068 | Address review feedback on v4 metadata format

Fix terminology and CDDL syntax issues identified in review:

- Remove "721-ERC-style" terminology; clarify reference to CIP-0025's
  nested map structure vs individual metadata fields
- Fix CDDL syntax: use proper bounded_bytes for "721", policy_id, and
  asset_name keys (matching CIP-0025 version 2 conventions)
- Remove __RESERVE_KEYWORD_721_V4__ marker; presence of "721" key is
  sufficient for format detection
- Update retrieval steps to detect format by checking for "721" key
- Clarify "backwards-compatible" in changelog: version 4 adds nested
  map format as additional option, compatible with v1-3 direct format
- Rename "721-style" example label to "nested map format"
- Propagate all fixes consistently across 222, 333, and 444 standards

thaddeusdiamond committed:
CIP-0068 | Add rationale for version 4 nested map format

Add comprehensive rationale explaining the motivation for introducing
the CIP-0025-compatible nested map format in version 4:

- Addresses minUTxO cost barriers for large NFT collections
- Enables metadata consolidation to reduce ADA locking (10-20% savings)
- Improves migration path from CIP-0025 to CIP-0068
- Preserves backward compatibility through strict optionality

This rationale explains the economic and practical benefits that led
to the version 4 design decision.

rphair reviewed (commented):
@thaddeusdiamond at the CIP meeting earlier today we spontaneously picked this one off the `Unconfirmed` list  and decided that we'd been looking for the response that you made a few hours later. 😅

In my opinion this satisfies all the demanding review points beginning with https://github.com/cardano-foundation/CIPs/pull/1112#issuecomment-3643025222 with corresponding changes in the text.

Personally I appreciate that https://github.com/cardano-foundation/CIPs/pull/1112#issuecomment-3642619340 was put literally into the _Rationale_ & I'd also like to see that happen for any remaining clarifications resulting from the debate — past, present or future — for the benefit of new implementors and others who may be interested in migrating their tokens.

@SmaugPool can you please go back & double-check the conversations beginning with your https://github.com/cardano-foundation/CIPs/pull/1112#discussion_r2611487924 and mark "resolved" the ones you believe have been properly addressed... or note otherwise if you think they need more debate or adjustment?

Otherwise I think we have a achieved a CIP candidate and so have put this on _Review_ for the next CIP meeting (https://hackmd.io/@cip-editors/126) & we'll double-check all pending review points again in 2 weeks.

@thaddeusdiamond @SmaugPool please also tag other associates who you think would be interested in confirming, disputing, or reviewing this update. 🙏

rphair reviewed (commented):
@thaddeusdiamond my guess is that @Ryun1 deleted that review (posted a few minutes before today's CIP meeting) after answering his own question about the backward compatibility (this had already been briefly settled at an earlier meeting).

We pressed this item through our Review agenda at that meeting and those present believed it to be technically in order, but with some concern coming from "whether wallet devs have asked for this".  The editors believe this could be an issue because of pathological cases like (hypothetically):
1. without interest, let alone buy-in, to this addition: this "next generation" of the CIP-0068 standard is merged in, but wallet implementations don't (or don't consistently) use it;
2. then some other conception of that "next generation" comes along and this update needs to be backed out somehow — or worse: somehow merged with another evolution that it might not be easily compatible with.

So our decision was to progress this if & when we can get some endorsement from tagged representatives from wallet development teams.  @Ryun1 and @perturbing said they would work on this and we're awaiting their tagging some of these for review.

In the meantime I think we can call this `Confirmed` (though not `Last Check` as it would be for an unquestioned update PR) as we would any technically complete update awaiting more detailed review.

Ryun1 commented:
@thaddeusdiamond appreciate the efforts 

It would be beneficial to see further review from a range of implementors
to confirm the motivation for this new version and that it is reasonable to be adopted

Tagging some known CIP-68 implementors for visibility
@AlexDochioiu @Scitz0 @refi93 @ashisherc

Scitz0 commented:
Ok, then it should be fairly easy to implement I think.

thaddeusdiamond commented:
@Scitz0 @rphair any updates on this CIP? Thanks.

Scitz0 commented:
> @Scitz0 @rphair any updates on this CIP? Thanks.

I haven't spent any time inplementing the change in Eternl but if its accepted and demand is there we will put it on the list of requested features.

rphair commented on CIP-0068/README.md:
```diff
@@ -155,16 +155,20 @@ The `user token` represents an NFT (non-fungible token).
+This is either:
+
+1. A low-level direct representation of the metadata, following closely the structure of CIP-0025
+2. A 721-ERC-style token mapping exactly matching the structure of CIP-0025
```
@thaddeusdiamond I'm resolving this as per your statement that you've 'removed the "721-ERC-style" terminology' which certainly seems to satisfy the review point.

rphair commented on CIP-0068/README.md:
```diff
@@ -304,6 +372,26 @@ metadata =
+      }
+    }
+  }
+
```
(also apparently fixed on one of the latter commits)

rphair reviewed (approved):
@thaddeusdiamond @Scitz0 (cc @SmaugPool @AlexDochioiu) I also agree that merging this will help confirm the demand for this method.  Please do whatever you can before & after an anticipated merge of this update to ensure developers know about the new specification.
* @Ryun1 @Crypto2099 please also post ASAP if you can think of any recommendations that should be made about compatibility issues (not likely I think, given the statement of "Full backward compatibility") or other developer announcements (and circulate these, since this mightn't be obvious from this PR's merge itself).

I think the last review about the details of nomenclature and CDDL have satisfied all significant community challenges so far & so I feel comfortable endorsing this and moving it on to `Last Check` for the upcoming meeting (https://hackmd.io/@cip-editors/134).
* Note we have many last-checks on this agenda due to a harmonic convergence of mature CIPs & CPSs... so although I'm happy to invite the author & all advocates of this proposal & we'll do our best to merge it on time, this can't be assured especially if other CIP editors still need to review it... but I'll ping @Ryun1 @perturbing to see if we can clear that requirement ahead of the meeting (with no rush of course).

thaddeusdiamond commented:
> @thaddeusdiamond @Scitz0 (cc @SmaugPool @AlexDochioiu) I also agree that merging this will help confirm the demand for this method. Please do whatever you can before & after an anticipated merge of this update to ensure developers know about the new specification.
> 
> * @Ryun1 @Crypto2099 please also post ASAP if you can think of any recommendations that should be made about compatibility issues (not likely I think, given the statement of "Full backward compatibility") or other developer announcements (and circulate these, since this mightn't be obvious from this PR's merge itself).
> 
> I think the last review about the details of nomenclature and CDDL have satisfied all significant community challenges so far & so I feel comfortable endorsing this and moving it on to `Last Check` for the upcoming meeting (https://hackmd.io/@cip-editors/134).
> 
> * Note we have many last-checks on this agenda due to a harmonic convergence of mature CIPs & CPSs... so although I'm happy to invite the author & all advocates of this proposal & we'll do our best to merge it on time, this can't be assured especially if other CIP editors still need to review it... but I'll ping @Ryun1 @perturbing to see if we can clear that requirement ahead of the meeting (with no rush of course).

Great! I will work on the implementation for my minting tools and look forward to seeing it live in Eternl.  In the meantime, since JPG.store is shutting down the only real implementers would be @SmaugPool, WayUp (I can reach out to Cash once it's merged) and the wallets.

thaddeusdiamond commented:
Sample consolidation transaction:
https://cardanoscan.io/transaction/6c900994bb705921cbcbdd40a404830b60997ea8015cca91bb4875d591a4f05c?tab=utxo

Sample consolidation UTxO datum:
```
d8799fa143373231a1581c8dcb1f87d74f3c973e6373aaa24f17ba9369e4af71bdc1973b37b732ae5157696c6454616e677a2053322036383032b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724843686172636f616c44626f64794744656661756c7444657965734653746f6e6564446e616d65581c57696c642054616e677a202d205365726965732032202d202339353145696d6167659f47697066733a2f2f583b6261667962656966737272376a706f3570333667356e3268676b66353273656472346f6f32743269736a7334793463347861646962373678767875ff456d6f7574684554656574684765796577656172474d6f6e6f636c6548636c6f7468696e674a562d4e65636b2054656548686561647765617257507572706c6520616e6420476f6c642042616c6c6361704a6261636b67726f756e644d466f756e7461696e20426c75654b6163636573736f726965734d47616d65646179205061696e744c757067726164656446726f6d4d57696c6454616e677a203935314e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322035303835b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a2053657269657320324366757244426c756544626f64794744656661756c7444657965734744656661756c74446e616d65581d57696c642054616e677a202d205365726965732032202d20233431323745696d6167659f47697066733a2f2f583b6261667962656968777362776c33337433793479687272686670706a367a7734366b697a61676269686e746b6163686b34326266633732786e7479ff456d6f75746845546565746847657965776561724841766961746f727348636c6f7468696e6752547572746c656e65636b2053776561746572486865616477656172464265616e69654a6261636b67726f756e644d41766f6361646f20477265656e4b6163636573736f7269657352536b656c65746f6e20466163657061696e744c757067726164656446726f6d4e57696c6454616e677a20343132374e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322035303533b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724843686172636f616c44626f64794744656661756c74446579657349426c61636b20457965446e616d65581d57696c642054616e677a202d205365726965732032202d20233338333645696d6167659f47697066733a2f2f583b626166796265696364636e696f3377777477793372326c33657a787962353469726f63667171706b6a3635726d6b6d35746872766a787368773661ff456d6f75746845426c756e744765796577656172474d6f6e6f636c6548636c6f7468696e6752547572746c656e65636b20537765617465724868656164776561724c4d4154412042616c6c6361704a6261636b67726f756e644a4f72616e6779205265644b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20333833364e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322035303635b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724542726f776e44626f64794744656661756c74446579657345416e677279446e616d65581d57696c642054616e677a202d205365726965732032202d20233131383645696d6167659f47697066733a2f2f583b6261667962656963326336346d3672777868633475357a7235783635353371786435656976757469647170646175787462636570616f6665706761ff456d6f7574684b467275697420536e61636b47657965776561724b536b6920476f67676c657348636c6f7468696e674854616e6720546f704868656164776561724742616e64616e614a6261636b67726f756e644d41766f6361646f20477265656e4b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20313138364e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034363030b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a2053657269657320324366757245477265656e44626f64794744656661756c74446579657349537175696e74696e67446e616d65581d57696c642054616e677a202d205365726965732032202d20233433313445696d6167659f47697066733a2f2f583b626166796265696876333634326e366f7a6d356e673670366378367167717367726864356c6c676a77367765326a6c6f62633467696b6c6a64356dff456d6f75746845546565746847657965776561724a53756e676c617373657348636c6f7468696e674d427573696e657373205375697448686561647765617247506c756e6765724a6261636b67726f756e6448426c7565204b6f694b6163636573736f726965734b416e67656c2057696e67734c757067726164656446726f6d4e57696c6454616e677a20343331344e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034363534b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724542726f776e44626f64794744656661756c7444657965734653746f6e6564446e616d65581d57696c642054616e677a202d205365726965732032202d20233430333345696d6167659f47697066733a2f2f583b62616679626569643779356c35636f72786e763437796f347a787677696e79766c626b7678643636663267636f6f74356b726a6b616e7061376a75ff456d6f7574684a476f6c6420546f6f746847657965776561724c5377696d20476f67676c657348636c6f7468696e6745436c6f616b486865616477656172444e6f6e654a6261636b67726f756e644a4f72616e6779205265644b6163636573736f72696573494e6f73652052696e674c757067726164656446726f6d4e57696c6454616e677a20343033334e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034353039b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a20536572696573203243667572464f72616e676544626f64794744656661756c74446579657349537175696e74696e67446e616d65581d57696c642054616e677a202d205365726965732032202d20233430303345696d6167659f47697066733a2f2f583b62616679626569646d6a376f617663366b73357671646e37653372646d6b796968656e646b6d6d33746c7072736a75676c787171616e376e6c3471ff456d6f75746845546565746847657965776561724841766961746f727348636c6f7468696e674e486177616969616e2053686972744868656164776561724842656572204861744a6261636b67726f756e644a4f72616e6779205265644b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20343030334e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034383735b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724843686172636f616c44626f64794744656661756c74446579657345416e677279446e616d65581d57696c642054616e677a202d205365726965732032202d20233338393445696d6167659f47697066733a2f2f583b626166796265696779726171766b366769697078346366623472716e6d63787962333673767a366965706167737978653568703761776368706769ff456d6f75746845546565746847657965776561724a53756e676c617373657348636c6f7468696e674e486177616969616e205368697274486865616477656172464265616e69654a6261636b67726f756e644d546f626163636f2042726f776e4b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20333839344e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034373831b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724843686172636f616c44626f64794744656661756c74446579657346437279696e67446e616d65581d57696c642054616e677a202d205365726965732032202d20233338303445696d6167659f47697066733a2f2f583b62616679626569633466366e767a326a61717135346d78723473326a366c71736272716c3273637578636f6c6d6e6b35796976376d366b77797375ff456d6f7574684642616e616e6147657965776561724a53756e676c617373657348636c6f7468696e674d427573696e6573732053756974486865616477656172484865616462616e644a6261636b67726f756e644f416d65746879737420507572706c654b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20333830344e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034373433b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a20536572696573203243667572464f72616e676544626f64794744656661756c7444657965734744656661756c74446e616d65581d57696c642054616e677a202d205365726965732032202d20233338303345696d6167659f47697066733a2f2f583b62616679626569636b6166676e78787a70676d3661676f6769326669773337716632327a677467346e736973347073706c327675646e6f75667575ff456d6f7574684943696761726574746547657965776561724c5377696d20476f67676c657348636c6f7468696e6752547572746c656e65636b205377656174657248686561647765617257507572706c6520616e6420476f6c642042616c6c6361704a6261636b67726f756e644d546f626163636f2042726f776e4b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20333830334e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034353438b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a205365726965732032436675724542726f776e44626f64794744656661756c74446579657345416e677279446e616d65581d57696c642054616e677a202d205365726965732032202d20233337373045696d6167659f47697066733a2f2f583b62616679626569633433327672616f6969326b377574656364646e7732656a36707964627a64766f76647a76626665676e6e716c6879646f327a34ff456d6f7574684553747261774765796577656172444e6f6e6548636c6f7468696e674c437265776e65636b205465654868656164776561724842656572204861744a6261636b67726f756e644c4361726d696e652050696e6b4b6163636573736f726965734b416e67656c2057696e67734c757067726164656446726f6d4e57696c6454616e677a20333737304e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034363634b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a2053657269657320324366757245477265656e44626f64794744656661756c74446579657345416e677279446e616d65581d57696c642054616e677a202d205365726965732032202d20233337313845696d6167659f47697066733a2f2f583b626166796265696466743468783671746261717972683373616d736a6364777562796d6474727463706d797632697a78626f766a7673767a757669ff456d6f7574684b467275697420536e61636b4765796577656172474d6f6e6f636c6548636c6f7468696e674d426f6d626572204a61636b6574486865616477656172484865616462616e644a6261636b67726f756e644d546f626163636f2042726f776e4b6163636573736f72696573494e6f73652052696e674c757067726164656446726f6d4e57696c6454616e677a20333731384e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034353830b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a2053657269657320324366757244426c756544626f64794744656661756c74446579657349426c61636b20457965446e616d65581d57696c642054616e677a202d205365726965732032202d20233336393145696d6167659f47697066733a2f2f583b626166796265696132793565346a7363746477786a77706c7632717266356c666968776f34773778666d766a3775626e6e7173706279646e6b7034ff456d6f75746845546565746847657965776561724a53756e676c617373657348636c6f7468696e674d426f6d626572204a61636b65744868656164776561724e436c6f7665722042616c6c6361704a6261636b67726f756e644d546f626163636f2042726f776e4b6163636573736f726965734d47616d65646179205061696e744c757067726164656446726f6d4e57696c6454616e677a20333639314e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e5157696c6454616e677a2053322034373335b0496d656469615479706549696d6167652f706e674770726f6a6563745357696c642054616e677a2053657269657320324366757244426c756544626f64794744656661756c74446579657349537175696e74696e67446e616d65581d57696c642054616e677a202d205365726965732032202d20233335393745696d6167659f47697066733a2f2f583b62616679626569686270376363347966746d366f75796e71756e6e65616c327a786c72717a786d6564766e716f676d68656d3735366b646a777a75ff456d6f7574684943696761726574746547657965776561724a53756e676c617373657348636c6f7468696e674e486177616969616e2053686972744868656164776561724a4865616470686f6e65734a6261636b67726f756e6448426c7565204b6f694b6163636573736f72696573444e6f6e654c757067726164656446726f6d4e57696c6454616e677a20333539374e6f726967696e616c506f6c69637958383333353638616431316639336233653739616538646565356164393238646564373261646365613731396539323130386361663135323162466172746973744b406e6574616e656c63686e04d8799f581c2d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c7581de19ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c56829ffff
```

Please use a datum inspector to validate the output JSON (e.g., https://cardanoscan.io/datumInspector)

The asset `asset143k4tan37qqx53u36an5smssd9yumvwc58zfxv` should render with this JSON from the CBOR into a well-formed NFT:

```
{
   "map_11":{
      "mapKey":"WildTangz S2 4664",
      "mapValue":{
         "map_0":{
            "mapKey":"mediaType",
            "mapValue":"image/png"
         },
         "map_1":{
            "mapKey":"project",
            "mapValue":"Wild Tangz Series 2"
         },
         "map_2":{
            "mapKey":"fur",
            "mapValue":"Green"
         },
         "map_3":{
            "mapKey":"body",
            "mapValue":"Default"
         },
         "map_4":{
            "mapKey":"eyes",
            "mapValue":"Angry"
         },
         "map_5":{
            "mapKey":"name",
            "mapValue":"Wild Tangz - Series 2 - #3718"
         },
         "map_6":{
            "mapKey":"image",
            "mapValue":[
               "ipfs://",
               "bafybeidft4hx6qtbaqyrh3samsjcdwubymdtrtcpmyv2izxbovjvsvzuvi"
            ]
         },
         "map_7":{
            "mapKey":"mouth",
            "mapValue":"Fruit Snack"
         },
         "map_8":{
            "mapKey":"eyewear",
            "mapValue":"Monocle"
         },
         "map_9":{
            "mapKey":"clothing",
            "mapValue":"Bomber Jacket"
         },
         "map_10":{
            "mapKey":"headwear",
            "mapValue":"Headband"
         },
         "map_11":{
            "mapKey":"background",
            "mapValue":"Tobacco Brown"
         },
         "map_12":{
            "mapKey":"accessories",
            "mapValue":"Nose Ring"
         },
         "map_13":{
            "mapKey":"upgradedFrom",
            "mapValue":"WildTangz 3718"
         },
         "map_14":{
            "mapKey":"originalPolicy",
            "mapValue":"33568ad11f93b3e79ae8dee5ad928ded72adcea719e92108caf1521b"
         },
         "map_15":{
            "mapKey":"artist",
            "mapValue":"@netanelchn"
         }
      }
   }
}
```

FYI @SmaugPool @Scitz0
